### ISSUE TITLE
Implementing some "default" auxiliary rules.

### ISSUE THREAD

**Issue by Maikuolan** [CIDRAM/CIDRAM#229]
I'm thinking about implementing some default auxiliary rules to CIDRAM (i.e., some basic auxiliary rules, which'll already exist in CIDRAM at the time when users install it, and which they can edit after having installed CIDRAM).

Since auxiliary rules are all handled by the one, same file (`auxiliary.yaml`), and since the updater never touches that file, this proposal, if implemented, would only affect *future* installations (i.e., *existing* installations would *not* suddenly have these rules appear, and would require the user to set them up themselves if they wanted them).

Some ideas:
- A default rule for enabling CAPTCHA for those profiles where enabling it might commonly be wanted (e.g., `Domestic ISP`). This also ties in with #228.
- A default rule for "redlisted" requests. Such requests would have all forms of CAPTCHA forcibly disabled, have the output template suppressed, and receive a 503 header. The rule could include some conditions which would never actually match anything in reality, but would serve as an example for users to be able to edit the rule as they see fit (e.g., CIDRs like `a.b.c.d/xx`, which isn't actually valid CIDR, but demonstrates the syntax and suggests obvious intentions). The idea of a "redlisted" request, is like a "blacklisted" request, but one which is considered so bad, so dangerous, or so unwanted, that we don't even want to waste a single byte more than necessary in dealing with that request, no second chances, like a "red alert" kind of request, if that makes sense. CIDRAM doesn't actually do this nor supported this directly in its own right, nor should it (since whether a request warrants a "red alert" is highly subjective anyway, and opinions will vary from one user to another), but it might be nice to provide an default rule to serve as an example of how to do this via auxiliary rules, for those that might want something like that.
- A default rule for "whitelisted" requests (e.g., "if your own IP address never changes, you can add it to this rule, to ensure that you'll never be blocked by your own installation", and that kind of thing, as a default rule to serve as an example for creating such whitelisting rules).

There are probably a ton more of ideas possible for where we could take this, but that should be sufficient to convey what I'm proposing here and to get the ball rolling. :-)

Please leave your comments here, or give a thumbs up/down, to let me know whether you like/dislike this proposal, to let me know whether you have any further ideas or thoughts about this, or ideas for more possible default rules.

Cheers.

**Comment by 737simpilot** [comment]
Sounds good. Since there's so much capability with the auxiliary rules I think it would be wise to document their abilities. 

When you mention red alert I think of this classic game I played back in the day. :grinning:


![tenor](https://user-images.githubusercontent.com/16517998/120315520-b0fbaa00-c299-11eb-8c01-38b70f15582a.gif)

**Comment by mikeruss1** [comment]
great idea, and a useful way for new users to understand the usefulness of rules.

needs clear step by step explanations - like a tutorial

**Comment by Maikuolan** [comment]
I worked on this a little this afternoon, and I think I've got something workable.

Would you guys mind taking a quick look, and letting me know what you reckon? You could always temporarily rename your own auxiliary rules to something else, and then upload this to your auxiliary rules file, or test it out locally, or on a separate installation or something like that, to see how it all looks at the auxiliary rules page. If you're up for it, let me know if you have any ideas for further improvements/additions, whether it's good as is, or whatever else. Cheers. :-)

```YAML
Whitelisted:
 Notes: >
  If your own IP address is static (i.e., never changes), you could consider utilising auxiliary rules to whitelist your
  own IP address, thus ensuring that it never gets blocked by your own installation. This rule here serves as an example,
  by whitelisting the fake IP address, "w.x.y.z" (you would replace this with your own, if you choose to use this rule).
  If you don't want to whitelist yourself, aren't able to, or aren't interested, you can simply delete this rule.
 Logic: "Any"
 Whitelist:
  If matches:
   IPAddr:
    - "w.x.y.z"
Redlisted:
 Notes: >
  This rule marks a request as being "redlisted". It utilises some fake, arbitrary values, which'll never actually match
  anything in the real world (unless you modify them to something valid). It's included just as an example only. The
  concept behind "redlisted" requests is similar to the concept behind "blacklisted" requests. The difference being that a
  "redlisted" request is one which is supposedly so bad, so dangerous, or so unwanted, that we don't want to waste even a
  single byte more than absolutely necessary in dealing with that request. When a "redlisted" request is blocked,
  reCAPTCHA and HCaptcha are forcibly disabled, output suppressed, IP tracking enabled, extended tracking options
  enforced, and a "503 Service Unavailable" header served. Because determining which requests should be "redlisted" is
  highly subjective, and CIDRAM aims to be as objective as possible when it comes to default rules or any standard
  assertions made, CIDRAM never makes such determinations itself (it'll be up to you, the user, to do that). If you're not
  interested in such a concept, you're welcome to simply delete this rule. Note that as this is a "profile" rule (and such
  rules are only intended to "profile" a request), a separate "block" rule *may* be needed to ensure that such requests
  are actually *blocked*.
 Logic: "Any"
 Forcibly disable reCAPTCHA: true
 Forcibly disable HCaptcha: true
 Suppress output template: true
 Forcibly enable IP tracking: true
 Enforce extended IP tracking options: true
 Profile:
  If matches:
   Factors:
    - "999.999.999.999/999"
    - "xxxx:yyyy::zzzz/999"
   UA:
    - "Example user agent for redlisting"
Non-strict:
 Notes: >
  When a "profile" is raised for a request, which identifies that request as a request where we may possibly want to
  enforce a less strict response than usual (e.g., ensuring that requests from domestic ISPs are provided with the
  opportunity to complete a CAPTCHA in order to regain access to your website), this rule ensures that an additional
  profile is raised: "Non-strict". This additional profile may then be optionally utilised as part of a chain of more
  complex auxiliary rules, intended to invoke various other more finely tuned, contextual, situational behaviours. Or
  alternatively, if you want CIDRAM to never enforce less strict responses, or if you have some other means of achieving
  that, you could, if you wanted to, simply delete this rule.
 Logic: "Any"
 Profile:
  If matches:
   Profile:
    - "Amateur Radio"
    - "Domestic ISP"
    - "Fixed Line ISP"
    - "Frequent changes"
    - "Government"
    - "Multiplay"
    - "Temporary"
    - "Tor endpoints here"
    - "University"
  But not if matches:
   Profile:
    - "Redlisted"
Mark non-strict for use with CAPTCHA:
 Notes: >
  For requests which raise the "non-strict" profile (e.g., requests from domestic ISPs), which trigger one or less
  signatures, and for whose IP address has incurred less than a thousand infractions, this rule will mark the request for
  use with reCAPTCHA and HCaptcha. In order for the rule to have any noticeable effect, CIDRAM must be configured
  accordingly (note, in particular, the "usemode" configuration directive available under the "recaptcha" and "hcaptcha"
  configuration categories). To reduce the risk of this rule interfering with bypasses or interfering with configuration
  values defined for reCAPTCHA/HCaptcha, the rule has been set as a "profile" rule (I strongly advise against changing the
  action associated with this rule). Modify or delete the rule according to your needs.
 Logic: "All"
 Mark for use with reCAPTCHA: true
 Mark for use with HCaptcha: true
 Profile:
  If matches:
   Profile:
    - "Non-strict"
   SignatureCount:
    - "<=1"
   Infractions:
    - "<1000"
Disable CAPTCHA for strict, moderately fractious requests:
 Notes: >
  As long as the "non-strict" profile hasn't been raised (thus excluding domestic ISPs, multiplay networks, etc), and as
  long as at least one signature has been triggered, for any requests whose infraction counts are equal to or greater than
  a thousand infractions, this rule will forcibly disable both reCAPTCHA and HCaptcha. Generally, such high infractions
  rates can be encountered as a result of "extended IP tracking options" being enforced, or by modules attempting to block
  detected hack attempts and similar such activity. To reduce the risk of this rule interfering with bypasses, the rule
  has been set as a "profile" rule, rather than as a "block" rule. Modify or delete the rule according to your needs.
 Logic: "Any"
 Forcibly disable reCAPTCHA: true
 Forcibly disable HCaptcha: true
 Profile:
  If matches:
   Infractions:
    - ">=1000"
   SignatureCount:
    - ">=1"
  But not if matches:
   Profile:
    - "Non-strict"
Suppress output for strict, highly fractious requests:
 Notes: >
  As long as the "non-strict" profile hasn't been raised (thus excluding domestic ISPs, multiplay networks, etc), and as
  long as at least one signature has been triggered, for any requests whose infraction counts are equal to or greater than
  3000 infractions, this rule will suppress output, will serve an additional block, and will serve a "503 Service
  Unavailable" header. Generally, such high infractions rates can be encountered as a result of "extended IP tracking
  options" being enforced, or by modules attempting to block detected hack attempts and similar such activity. By the
  point that such a rule is triggered, there's typically very little chance that we'd ever want the request which
  triggered it to be able to get through, or that we'd want to be able to perform additional processing to invoke any
  further, more complex behaviour, so the rule has been set as a "block" rule accordingly, rather than as a "profile" rule
  or otherwise. Modify or delete the rule according to your needs.
 Logic: "All"
 Reason: "Request too fractious"
 Status Code: "503"
 Suppress output template: true
 Block:
  If matches:
   Infractions:
    - ">=3000"
   SignatureCount:
    - ">=1"
  But not if matches:
   Profile:
    - "Non-strict"
Using wildcards to block unwanted traffic:
 Method: "WinEx"
 Notes: "This rule serves as an example. It utilises Windows-style wildcards to block any requests with a matching user agent."
 Logic: "Any"
 Reason: "Example auxiliary rule"
 Block:
  If matches:
   UA:
    - "As long as I'm on the left, anything can be on the right*"
    - "*As long as I'm on the right, anything can be on the left"
Using regular expressions to block unwanted traffic:
 Method: "RegEx"
 Notes: >
  This rule serves as an example. It utilises regular expressions to block any requests with a matching user agent (the
  example looks for any version of the imaginary "Foo Bar" user agent; e.g., "Foo Bar v1.234.567 (+http://foo.bar)").
 Logic: "Any"
 Reason: "Example auxiliary rule"
 Block:
  If matches:
   UA:
    - "~^Foo Bar v\d+\.\d+\.\d+ \(+http://foo.bar\)$~"
```

**Comment by mikeruss1** [comment]
I have a feeling that what I have to say is not going to be helpful ....
I think its too complex too quickly, uses too much pre-existing knowledge, and would be tough for an average new user.
suggest start off simple, perhaps say you want to block a particular UA - this is how you do it,
then add to that with say you want to not offer recaptcha when this happens - OK do this
then not track it, or not report it
then introduce profile, simple example of a profile and how to use it
and then gradually build more complex cases.

**Comment by Maikuolan** [comment]
I wouldn't call it unhelpful at all. Actually, I think you've got a good point there. In particular, looking over it all again myself, I'm thinking I've probably got a fair bit more text there already than most users would want to deal with, at least as first timers.

Maybe I need to keep the default rules much more simple, but then just compliment it with some YouTube tutorials or something like that? Probably easier to demonstrate how to perform specific tasks for first timers using video tutorials than it would be using default rules.

**Comment by mikeruss1** [comment]
that will work I reckon.

**Comment by Maikuolan** [comment]
Revision:

```YAML
Non-strict:
 Notes: >
  Triggers when other signatures, where we may want the response to be less strict than usual (e.g., when blocking
  domestic ISPs), have already been triggered. Profiles the request as "non-strict".
 Logic: "Any"
 Profile:
  If matches:
   Profile:
    - "Amateur Radio"
    - "Domestic ISP"
    - "Fixed Line ISP"
    - "Frequent changes"
    - "Government"
    - "Multiplay"
    - "Temporary"
    - "Tor endpoints here"
    - "University"
  But not if matches:
   Profile:
    - "Blacklisted"
    - "Redlisted"
Mark non-strict for use with CAPTCHA:
 Notes: >
  Triggers when the request has been profiled as "non-strict", when one or less signatures have been triggered, and the
  request's IP address has incurred less than a thousand infractions. Marks the request for use with reCAPTCHA and
  HCaptcha. In order for the rule to have any noticeable effect, CIDRAM must be configured accordingly (e.g., the
  "usemode" configuration directive).
 Logic: "All"
 Mark for use with reCAPTCHA: true
 Mark for use with HCaptcha: true
 Profile:
  If matches:
   Profile:
    - "Non-strict"
   SignatureCount:
    - "<=1"
   Infractions:
    - "<1000"
```

This just covers the behaviours removed by #228 (so that we can rely on auxiliary rules to "mark" things, instead of needing to rely on the signature files). In this revision, no "example" rules are provided, and the text/explanation is kept very minimal.

I haven't had the time (or much motivation) yet to create any video tutorials or similar, but I'm thinking I can just create a separate issue for that (seeing as tutorials wouldn't strictly be related to anything "default").

In terms of just the "default auxiliary rules", reckon that should be enough..?

**Comment by 737simpilot** [comment]
I think it would be wise to consider people are either a visual learner or an audio learner. Though, now that I remember, a while back I watched a pretty informative YouTube video on this learning process thing and it may not be true. LOL Here's the video: https://www.youtube.com/watch?v=rhgwIhB58PA Do check it out when you have the time. He's got some great stuff there on his channel. (14 minutes). 

I just saw a program from my favorite software website, snapfiles.com for creating news letters, business cards, and what not. That could perhaps be used in lieu of Microsoft software for documentation.  Though, one reviewer of the program indicated it wasn't a very well polished program. It's called Scribus. `https://www.snapfiles.com/get/scribus.html`

Or everyone's favorite PowerPoint comes to mind. LibreOffice calls theirs Impress. 

I have never used Impress or PowerPoint, but I've been interested in learning as of late. Using Impress or PowerPoint could allow one to use chapters _I think_ to quickly go to the pertinent Info. for all of CIDRAM's documentation I think. It would of course take lots of time to accomplish. Though, I'd like to do it, but I'm not comfortable as I'm not the Dev and may not know the finer inner workings that should be described as well as other things. I could use the current documentation and this posting above about the auxiliary rules to aid in its creation though. I have been told I had a nice voice.. LOL! So adding audio as well as visual information to the PowerPoint/Impress presentation would be what I'd do. 

Anyway...



![img](https://i.postimg.cc/JhbLkmSZ/ezgif-com-gif-maker-5.gif)
